File: app/src/main/java/com/wang17/myphone/activity/SmsActivity.java
package com.wang17.myphone.activity;

import android.content.DialogInterface;
import android.content.Intent;
import android.database.Cursor;
import android.graphics.Color;
import android.net.Uri;
import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AlertDialog;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.util.Log;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.BaseAdapter;
import android.widget.EditText;
import android.widget.LinearLayout;
import android.widget.ListView;
import android.widget.TextView;

import com.wang17.myphone.R;
import com.wang17.myphone.model.DateTime;
import com.wang17.myphone.database.PhoneMessage;
import com.wang17.myphone.database.DataContext;
import com.wang17.myphone.util._Utils;
import com.wang17.myphone.widget.MyWidgetProvider;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.regex.Matcher;
import java.util.regex.Pattern;

public class SmsActivity extends AppCompatActivity {

    private List<PhoneMessage> mSmsList;
    private SmsListdAdapter adapter;
    private ListView mSmsListView;
    private FloatingActionButton actionButtonAdd;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_sms);

        mSmsList = new ArrayList<>();
        adapter = new SmsListdAdapter();
        mSmsListView = findViewById(R.id.listView_sms);

        // 添加异常捕捉
        try {
            Cursor cursor = getContentResolver().query(
                    Uri.parse("content://sms"),
                    new String[]{"_id", "address", "read", "body", "date"},null, null, "date desc"); // datephone想要的短信号码

            if (cursor != null) { // 当接受到的新短信与想要的短信做相应判断

                String body = "";
                while (cursor.moveToNext()) {
                    body = cursor.getString(cursor.getColumnIndex("body"));// 在这里获取短信信息
                    long smsdate = Long.parseLong(cursor.getString(cursor.getColumnIndex("date")));
                    e(body);
                }

            }
            DataContext dataContext = new DataContext(this);
            mSmsList = dataContext.getPhoneMessages();
            mSmsListView.setAdapter(adapter);
//            }
        } catch (Exception ex) {
            _Utils.printException(this, ex);
            Log.e("wangsc", ex.getMessage());
        }
    }

    private void e(Object log) {
        Log.e("wangsc",log.toString());
    }

    class SmsModel {
        String number;
        String body;
        DateTime smsDate;

        public SmsModel(String number, String body, DateTime smsDate) {
            this.number = number;
            this.body = body;
            this.smsDate = smsDate;
        }
    }

    private int preDay;
    private HashMap<Integer, View> mHashMap = new HashMap<>();

    protected class SmsListdAdapter extends BaseAdapter {
        @Override
        public int getCount() {
            return mSmsList.size();
        }

        @Override
        public Object getItem(int position) {
            return null;
        }

        @Override
        public long getItemId(int position) {
            return 0;
        }

        @Override
        public View getView(int position, View convertView, ViewGroup parent) {
            try {

                convertView = View.inflate(SmsActivity.this, R.layout.inflate_list_item_sms, null);

                if (mHashMap.get(position) == null) {
                    mHashMap.put(position, convertView);

                    PhoneMessage sms = mSmsList.get(position);
                    LinearLayout layoutDate = convertView.findViewById(R.id.layout_date);
                    LinearLayout root = convertView.findViewById(R.id.linearLayout_root);
                    TextView textViewDate = convertView.findViewById(R.id.tv_att_date);
                    TextView textViewDateTime = convertView.findViewById(R.id.textView_dateTime);
                    TextView textViewBody = convertView.findViewById(R.id.textView_body);
                    TextView textViewNumber = convertView.findViewById(R.id.textView_number);
                    root.setOnLongClickListener(new View.OnLongClickListener() {
                        @Override
                        public boolean onLongClick(View v) {

                            new AlertDialog.Builder(SmsActivity.this).setMessage("要删除所有短信信息吗？").setPositiveButton("是", new DialogInterface.OnClickListener() {
                                @Override
                                public void onClick(DialogInterface dialog, int which) {
                                    DataContext dataContext = new DataContext(SmsActivity.this);
                                    dataContext.deletePhoneMessage();
                                    mSmsList = dataContext.getPhoneMessages();
                                    adapter.notifyDataSetChanged();
                                }
                            }).show();
                            return true;
                        }
                    });
                    if (sms.getCreateTime().getDay() != preDay) {
                        textViewDate.setText(sms.getCreateTime().toShortDateString());
                        layoutDate.setVisibility(View.VISIBLE);
                        preDay = sms.getCreateTime().getDay();
                    } else {
                        layoutDate.setVisibility(View.GONE);
                    }
                    textViewDateTime.setText(sms.getCreateTime().toLongDateTimeString());
                    textViewBody.setText(sms.getBody());
                    textViewNumber.setText(sms.getAddress());

                    Matcher matcher = Pattern.compile("-?\\d+\\.\\d+").matcher(sms.getBody());
                    if (matcher.find()) {
                        double money = Double.parseDouble(matcher.group());
                        if (Math.abs(money) > 100) {
                            textViewBody.setTextColor(Color.BLACK);
                        }
                    }

                } else {
                    convertView = mHashMap.get(position);
                }
            } catch (Exception e) {
                _Utils.printException(SmsActivity.this, e);
            }
            return convertView;
        }
    }


    private void delTodoDialog() {
        final View view = LayoutInflater.from(this).inflate(R.layout.dialog_del_todo, null);

        new AlertDialog.Builder(this)
                .setView(view).setCancelable(false).setPositiveButton("OK", new DialogInterface.OnClickListener() {
            @Override
            public void onClick(DialogInterface dialog, int which) {

                try {
                    EditText editTextTitle = view.findViewById(R.id.editText_title);
                    EditText editTextSummary = view.findViewById(R.id.editText_summary);

                    String bankName = editTextTitle.getText().toString();
                    String cardNumber = editTextSummary.getText().toString();
                    DataContext dataContext = new DataContext(SmsActivity.this);

                    dataContext.deleteBankToDo(bankName, cardNumber);

                    // 发送更新广播
                    Intent intent = new Intent(MyWidgetProvider.ACTION_UPDATE_LISTVIEW);
                    sendBroadcast(intent);

                } catch (Exception e) {
                    _Utils.printException(SmsActivity.this, e);
                }
            }
        }).setNegativeButton("CANCEL", null).create().show();
    }

}

File: app/src/main/java/com/wang17/myphone/activity/SmsActivity.kt
package com.wang17.myphone.activity

import android.annotation.SuppressLint
import android.content.DialogInterface
import android.os.Bundle
import android.support.v7.app.AlertDialog
import android.support.v7.app.AppCompatActivity
import android.util.Log
import android.view.View
import android.view.ViewGroup
import android.widget.*
import com.wang17.myphone.R
import com.wang17.myphone.dao.DataContext
import com.wang17.myphone.database.PhoneMessage
import com.wang17.myphone.model.DateTime
import com.wang17.myphone.util._Utils.printException
import kotlinx.android.synthetic.main.activity_sms.*
import kotlinx.android.synthetic.main.widget_timer.*
import java.util.*
import java.util.regex.Pattern
import kotlin.collections.ArrayList
import kotlin.collections.HashMap


class SmsActivity : AppCompatActivity() {


    class GroupItem(
        var dateTime: DateTime,
        var number: String,
        var name: String,
        var body: String
    )

    class ChildItem(
        var id: UUID,
        var number: String,
        var dateTime: DateTime,
        var body: String
    )

    private var groupList: MutableList<GroupItem> = ArrayList()
    private var childListList: MutableList<MutableList<ChildItem>> = ArrayList()
    private var smsList: List<PhoneMessage> = ArrayList()
    private var expandAdapter: SmsExpandableListAdapter = SmsExpandableListAdapter()
    private lateinit var dc: DataContext

    @SuppressLint("RestrictedApi")
    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_sms)

        try {
            dc = DataContext(this)
            loadSmsData()

            elv_sms.setGroupIndicator(null)
            elv_sms.setAdapter(expandAdapter)

            elv_sms.setOnChildClickListener { parent, v, groupPosition, childPosition, id ->
                AlertDialog.Builder(this).setMessage("确认删除?").setPositiveButton("删除", DialogInterface.OnClickListener { dialog, which ->
                    val ss = childListList[groupPosition][childPosition]
                    dc.deletePhoneMessage(ss.id)
                    loadSmsData()
                    expandAdapter.notifyDataSetChanged()
                }).show()
                true
            }



            fab_clean.setOnClickListener {
                val removeList = smsList.filter { m ->
                    !m.body.contains("银行") && !m.body.contains("信用卡") && !m.body.contains("金额") && !m.body.contains("余额") && !m.body.contains("账号") && !m.body.contains("账户")
                            && !m.body.contains("转账") && !m.body.contains("转出") && !m.body.contains("转入") && !m.body.contains("账单")
                            && !m.body.contains("还款") && !m.body.contains("借款") && !m.body.contains("贷款")
                            && !m.body.contains("快递") && !m.body.contains("物流")
                            || m.body.contains("验证码") || m.body.contains("动态密码")
                }
                dc.deletePhoneMessage(removeList)
                loadSmsData()
                expandAdapter.notifyDataSetChanged()
                AlertDialog.Builder(this).setMessage("整理完毕！").show()
            }
            fab_clean.setOnLongClickListener {
                dc.deletePhoneMessage()
                loadSmsData()
                expandAdapter.notifyDataSetChanged()
                AlertDialog.Builder(this).setMessage("清空完毕！").show()

                true
            }

            fab_clean_history.setOnClickListener {
                dc.deletePhoneMessageFrom(System.currentTimeMillis() - 1000 * 60 * 60 * 24 * 20L)
                loadSmsData()
                expandAdapter.notifyDataSetChanged()
                AlertDialog.Builder(this).setMessage("删除完毕！").show()
            }
            fab_express.setOnClickListener {
                val removeList = smsList.filter { m ->
                    m.body.contains("快递")
                }
                dc.deletePhoneMessage(removeList)
                loadSmsData()
                expandAdapter.notifyDataSetChanged()
                AlertDialog.Builder(this).setMessage("删除完毕！").show()
            }
            //            }
        } catch (ex: Exception) {
            printException(this, ex)
            Log.e("wangsc", ex.message!!)
        }
    }

    fun findName(body: String): String? {
        var matcher = Pattern.compile("(?<=【).{1,10}(?=】)").matcher(body)
        if (matcher.find())
            return matcher.group().trim()

        matcher = Pattern.compile("(?<=\\[).{1,10}(?=\\])").matcher(body)
        if (matcher.find())
            return matcher.group().trim()

        return null
    }

    fun loadSmsData() {
        groupList.clear()
        childListList.clear()
        smsList = dc.phoneMessages
        smsList.forEach { sms ->
            val group = groupList.firstOrNull { m -> m.number == sms.address }
            var index = groupList.indexOfFirst { m -> m.number == sms.address }
            if (group == null) {
                groupList.add(GroupItem(sms.createTime, sms.address, findName(sms.body) ?: sms.address, sms.body))
                var childList = ArrayList<ChildItem>()
                childList.add(ChildItem(sms.id, sms.address, sms.createTime, sms.body))
                childListList.add(childList)
            } else {
//                if ( sms.createTime.timeInMillis>group.dateTime.timeInMillis) {
//                    group.dateTime = sms.createTime
//                    group.body = sms.body
//                }
                findName(sms.body)?.let {
                    group.name = it
                }
                childListList.get(index).add(ChildItem(sms.id, sms.address, sms.createTime, sms.body))
            }
        }
    }

    private fun e(log: Any) {
        Log.e("wangsc", log.toString())
    }

    internal inner class SmsModel(var number: String, var body: String, var smsDate: DateTime)

    private var preDay = 0
    private val mHashMap = HashMap<Int, View?>()


    protected inner class SmsExpandableListAdapter : BaseExpandableListAdapter() {
        override fun getGroupCount(): Int {
            return groupList.size
        }

        override fun getChildrenCount(groupPosition: Int): Int {
            return childListList[groupPosition].size
        }

        override fun getGroup(groupPosition: Int): Any {
            return groupList.get(groupPosition)
        }

        override fun getChild(groupPosition: Int, childPosition: Int): Any {
            return childListList.get(groupPosition).get(childPosition)
        }

        override fun getGroupId(groupPosition: Int): Long {
            return groupPosition.toLong()
        }

        override fun getChildId(groupPosition: Int, childPosition: Int): Long {
            return childPosition.toLong()
        }

        override fun hasStableIds(): Boolean {
            return true
        }

        override fun getGroupView(groupPosition: Int, isExpanded: Boolean, convertView: View?, parent: ViewGroup?): View? {
            var view = convertView
            try {
                view = View.inflate(this@SmsActivity, R.layout.inflate_list_sms_group, null)
                val tvDateTime = view.findViewById<TextView>(R.id.tv_dateTime)
                val tvNumber = view.findViewById<TextView>(R.id.tv_number)
                val tvBody = view.findViewById<TextView>(R.id.tv_body)
                val groupItem = groupList.get(groupPosition)

                val today = DateTime.today
                val yestoday = today.addDays(-1)

                if (groupItem.dateTime.timeInMillis < yestoday.timeInMillis) {
                    tvDateTime.text = groupItem.dateTime.toShortDateString1()
                } else if (groupItem.dateTime.timeInMillis < today.timeInMillis) {
                    tvDateTime.text = "昨天"
                } else if (groupItem.dateTime.timeInMillis >= today.timeInMillis) {
                    tvDateTime.text = groupItem.dateTime.toShortTimeString()
                }
                tvNumber.text = groupItem.name
                tvBody.text = groupItem.body
            } catch (ex: Exception) {
            }
            return view
        }

        override fun getChildView(groupPosition: Int, childPosition: Int, isLastChild: Boolean, convertView: View?, parent: ViewGroup?): View? {
            var view = convertView
            try {
                view = View.inflate(this@SmsActivity, R.layout.inflate_list_sms_child, null)
                val tvDateTime = view.findViewById<TextView>(R.id.tv_dateTime)
                val tvBody = view.findViewById<TextView>(R.id.tv_body)
                val item = childListList.get(groupPosition).get(childPosition)

                val today = DateTime.today
                val yestoday = today.addDays(-1)

                if (item.dateTime.timeInMillis < yestoday.timeInMillis) {
                    tvDateTime.text = item.dateTime.toShortDateString1()
                } else if (item.dateTime.timeInMillis < today.timeInMillis) {
                    tvDateTime.text = "昨天"
                } else if (item.dateTime.timeInMillis >= today.timeInMillis) {
                    tvDateTime.text = item.dateTime.toShortTimeString()
                }
                tvBody.text = item.body
            } catch (ex: Exception) {
            }
            return view
        }

        override fun isChildSelectable(groupPosition: Int, childPosition: Int): Boolean {
            return true
        }

    }

}